### PR TITLE
Add unit tests for match function in analytics controller

### PR DESCRIPTION
**Pull request by alierkilic**
This will address issue: Unit tests for crane #91

If you have any problems with the test names please let me know

### CODE REVIEW

**Comment by qmhu** [comment]
@alierkilic thanks for you contribution, you can combine your unit test into one like :https://github.com/gocrane/crane/blob/main/pkg/controller/ehpa/hpa_test.go#L6

**Comment by alierkilic** [comment]
Should I mock the function calls inside Reconcile, SetupWithManager and createRecommendation? I don't see gomock used in the project.

**Comment by mfanjie** [comment]
@alierkilic mock testing are wanted! you don't see those tests because we are still focusing on core feature development. It's perfect if you can start from such tasks.

**Comment by mfanjie** [comment]
can you squash commits of this PR?

**Comment by alierkilic** [comment]
@mfanjie I will start working on it

**Comment by mfanjie** [comment]
@alierkilic you can still refine this pr and merge the commit, more UT can be done in separated PR.

**Comment by alierkilic** [comment]
@mfanjie can you accept this PR? If the merge commit is a problem, I can force push an older head, merge and then add my changes. It was like that before. I think I accidentally closed it.

**Comment by mfanjie** [comment]
plz squash commits

File: pkg/controller/analytics/analytics_controller_test.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+	unmatchingLabels["key2"] = "value2"
+
+	tests := []struct {
+		description   string
+		matchLabels   map[string]string
+		labelSelector metav1.LabelSelector
+		expect        bool
+	}{
+		{
+			description: "unmatching match labels and label selector match labels",
```
Comment (mfanjie) [inline pkg/controller/analytics/analytics_controller_test.go:23]: description: "match labels",

File: pkg/controller/analytics/analytics_controller_test.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
+package analytics
+
+import (
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestMatch(t *testing.T) {
+	matchingLabels := make(map[string]string)
```
Comment (mfanjie) [inline pkg/controller/analytics/analytics_controller_test.go:10]: matchingLabels -> matchLabels

File: pkg/controller/analytics/analytics_controller_test.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	matchingLabels["key1"] = "value1"
+
+	unmatchingLabels := make(map[string]string)
```
Comment (mfanjie) [inline pkg/controller/analytics/analytics_controller_test.go:13]: unmatchLabels := make(map[string]string)

File: pkg/controller/analytics/analytics_controller_test.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+			description: "unmatching match labels and label selector match labels",
+			matchLabels: matchingLabels,
+			labelSelector: metav1.LabelSelector{
+				MatchLabels:      unmatchingLabels,
+				MatchExpressions: []metav1.LabelSelectorRequirement{},
+			},
+			expect: false,
+		},
+		{
+			description: "matching match labels and label selector with operator 'Exists' and unmatching key",
```
Comment (mfanjie) [inline pkg/controller/analytics/analytics_controller_test.go:32]: description: "match expression key2 exists"

File: pkg/controller/analytics/analytics_controller_test.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+				MatchExpressions: []metav1.LabelSelectorRequirement{{
+					Key:      "key2",
+					Operator: "Exists",
+					Values:   []string{},
+				}},
+			},
+			expect: false,
+		},
+		{
+			description: "matching match labels and label selector with operator 'Exists' and matching key",
```
Comment (mfanjie) [inline pkg/controller/analytics/analytics_controller_test.go:45]: description: "match expression key1 exists"

File: pkg/controller/analytics/analytics_controller_test.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+				MatchExpressions: []metav1.LabelSelectorRequirement{{
+					Key:      "key1",
+					Operator: "Exists",
+					Values:   []string{},
+				}},
+			},
+			expect: true,
+		},
+		{
+			description: "matching match labels and label selector with operator 'DoesNotExist' and matching key",
```
Comment (mfanjie) [inline pkg/controller/analytics/analytics_controller_test.go:58]: description: "match expression key1 doesNotExists"

File: pkg/controller/analytics/analytics_controller_test.go [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+				MatchExpressions: []metav1.LabelSelectorRequirement{{
+					Key:      "key1",
+					Operator: "DoesNotExist",
+					Values:   []string{},
+				}},
+			},
+			expect: false,
+		},
+		{
+			description: "matching match labels and label selector with operator 'DoesNotExist' and unmatching key",
```
Comment (mfanjie) [inline pkg/controller/analytics/analytics_controller_test.go:71]: description: "match expression key2 doesNotExists"

File: pkg/controller/analytics/analytics_controller_test.go [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+				MatchExpressions: []metav1.LabelSelectorRequirement{{
+					Key:      "key2",
+					Operator: "DoesNotExist",
+					Values:   []string{},
+				}},
+			},
+			expect: true,
+		},
+		{
+			description: "matching match labels and label selector with operator 'In' and unmatching key",
```
Comment (mfanjie) [inline pkg/controller/analytics/analytics_controller_test.go:84]: description: "match expression key2 in"

File: pkg/controller/analytics/analytics_controller_test.go [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+				MatchExpressions: []metav1.LabelSelectorRequirement{{
+					Key:      "key2",
+					Operator: "In",
+					Values:   []string{},
+				}},
+			},
+			expect: false,
+		},
+		{
+			description: "matching match labels and label selector with operator 'In', matching key and matching value",
```
Comment (mfanjie) [inline pkg/controller/analytics/analytics_controller_test.go:97]: description: "match expression key1 in value1"

File: pkg/controller/analytics/analytics_controller_test.go [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+				MatchExpressions: []metav1.LabelSelectorRequirement{{
+					Key:      "key1",
+					Operator: "In",
+					Values:   []string{"value1"},
+				}},
+			},
+			expect: true,
+		},
+		{
+			description: "matching match labels and label selector with operator 'In', matching key and unmatching value",
```
Comment (mfanjie) [inline pkg/controller/analytics/analytics_controller_test.go:110]: description: "match expression key1 in value2"

File: pkg/controller/analytics/analytics_controller_test.go [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+				MatchExpressions: []metav1.LabelSelectorRequirement{{
+					Key:      "key1",
+					Operator: "In",
+					Values:   []string{"value2"},
+				}},
+			},
+			expect: false,
+		},
+		{
+			description: "matching match labels and label selector with operator 'NotIn', matching key and matching value",
```
Comment (mfanjie) [inline pkg/controller/analytics/analytics_controller_test.go:123]: description: "match expression key1 notIn value1"

File: pkg/controller/analytics/analytics_controller_test.go [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+				MatchExpressions: []metav1.LabelSelectorRequirement{{
+					Key:      "key1",
+					Operator: "NotIn",
+					Values:   []string{"value1"},
+				}},
+			},
+			expect: false,
+		},
+		{
+			description: "matching match labels and label selector with operator 'NotIn', matching key and unmatching value",
```
Comment (mfanjie) [inline pkg/controller/analytics/analytics_controller_test.go:136]: description: "match expression key1 notIn Value2"

**Review by mfanjie — COMMENTED** [review]
suggest to simplify the description

**Comment by alierkilic** [comment]
> suggest to simplify the description

done

**Pull request merged**